### PR TITLE
Parallel and more robust zbwardrive

### PR DESCRIPTION
**Pull request by rfmcpherson**
Major Changes:
Parallel wardriving using multiprocessing over many devices
More robust to api-mote timeouts and ravenstick crashes
zbwardrive logging
Reading and scanning timeouts
Support for "fake gps"

### CODE REVIEW

**Comment by rfmcpherson** [comment]
I've made some minor changes since I lost access to a working gps. The code should work, but it would be a good idea for someone with a working GPS to check it.

File: killerbee/pcapdump.py [hunk 1]
```diff
@@ -103,27 +106,24 @@ def pcap_next(self):
...
-            raise ValueError("Unsupported type for 'savefile' argument")
-
+        self.ppi = ppi        
+        if folder[-1] == "/":
+            folder = folder[:-1]
+        if not os.path.exists(folder+"/pcap"):
+            log_message = "PcapDump: Creating directory to store results"
+            logging.debug(log_message)
+            os.makedirs(folder+"/pcap")
+        self.__fh = open(folder+"/pcap"+savefile, mode='wb')
```
Comment (rmspeers) [inline killerbee/pcapdump.py]: The way I read this, it assumes that the user must always want the files in a /pcap folder either in their CWD by default or in another path they pass. If a tool wants to use such a restriction, that seems fine, but I think that the output serialization functions should accept whatever output path (savefile parameter) that the user wants - as people may want to do other things than save to SOMETHING/pcap/SOMETHING.

File: killerbee/pcapdump.py [hunk 2]
```diff
@@ -203,8 +197,10 @@ def pcap_dump(self, packet, ts_sec=None, ts_usec=None, orig_len=None,
                     raise Exception("Longitude value is out of expected range: %.8f" % lon)
                 if alt > -180000.00005 and alt < 180000.00005:
                     alt_i = int(round((alt + 180000.0) * 1e4))
+                elif math.isnan(alt):
+                    alt_i = 4294967295
```
Comment (rmspeers) [inline killerbee/pcapdump.py]: What is this special value and when does it play in?

File: killerbee/zbwardrive/findgps.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+# Helpful script to find where the gps is
+
+for sysdevpath in $(find /sys/bus/usb/devices/usb*/ -name dev); do
+    (
+        syspath="${sysdevpath%/dev}"
+        devname="$(udevadm info -q name -p $syspath)"
+        eval "$(udevadm info -q property --export -p $syspath)"
+	if [[ "$ID_SERIAL" = *"Prolific"* ]]
```
Comment (rmspeers) [inline killerbee/zbwardrive/findgps.sh]: What about for other vendors? Maybe add in the comments that this is only tested so far with device ___ version ___ and people are encouraged to add support for other devices, etc?

File: killerbee/pcapdump.py (hunk 1 above)
Comment (rfmcpherson) [inline killerbee/pcapdump.py]: Sounds good. I'll switch that

File: killerbee/pcapdump.py (hunk 2 above)
Comment (rfmcpherson) [inline killerbee/pcapdump.py]: I didn't write this part. I'm assuming it's a way to catch if alt hasn't been set by the GPS and give it a huge number (2**32-1) to show that it's unknown.

File: killerbee/zbwardrive/findgps.sh (hunk 3 above)
Comment (rfmcpherson) [inline killerbee/zbwardrive/findgps.sh]: Sounds good. Will do.

**Comment by rfmcpherson** [comment]
Been waylaid with a bunch of stuff at work. I hope to complete this in December.

**Comment by rmspeers** [comment]
Hey -- did you have any chance to make those tweaks? Would love to finish up this PR and merge it in.

**Comment by rfmcpherson** [comment]
Sorry. This completely slipped my mind. I'll add this to my todo list and finish it up shortly.

File: killerbee/zbwardrive/scanning.py [hunk 4]
```diff
@@ -1,147 +1,337 @@
...
+    # Iterate through all the Scanners and see if they died
+    # Respawn them if they have.
+    # (The Ravensticks occasionally suffer glib errors)
+    try:
+        while 1:
+            for i, s in enumerate(scanners):
+                
+                # Wait on the join and then start it again if it died
+                s["proc"].join(1)
+                if not s["proc"].is_alive():
```
Comment (rmspeers) [inline killerbee/zbwardrive/scanning.py]: How do you detect a dead rzusbstick? Another contributor's code just resets every 25 messages but I'd love to see that updated to a logic check.

Comment (rfmcpherson) [inline killerbee/zbwardrive/scanning.py]: Check out create_device on line 128. It fires an alarm after 10 seconds if nothing happens and then tries again. If it fails to work 5 times in a row, it throws an exception. In my experience (and from what I gathered talking to you last summer) they rarely fail more than 2-3 times in a row.

**Comment by rfmcpherson** [comment]
Changes based off of comments and more gps testing.

File: killerbee/pcapdump.py [hunk 5]
```diff
@@ -103,27 +106,24 @@ def pcap_next(self):
...
-            raise ValueError("Unsupported type for 'savefile' argument")
-
+        self.ppi = ppi        
+        if folder[-1] == "/":
+            folder = folder[:-1]
+        if not os.path.exists(folder):
+            log_message = "PcapDump: Creating directory to store results"
+            logging.debug(log_message)
+            os.makedirs(folder)
+        self.__fh = open(folder+savefile, mode='wb')
```
Comment (rmspeers) [inline killerbee/pcapdump.py]: As pcapdump I think is used across a few applications, I can't have this write files all the time. If we want it to do a write of this way, it needs to default to not happening and be a parameter that turns it on.

File: killerbee/zbwardrive/findgps.sh (hunk 3 above)
Comment (rmspeers) [inline killerbee/zbwardrive/findgps.sh]: Still a low priority thing, but one that we should add before merging for code quality.

File: killerbee/pcapdump.py (hunk 5 above)
Comment (rfmcpherson) [inline killerbee/pcapdump.py]: Will do.

File: killerbee/zbwardrive/zbwardrive.py [hunk 6]
```diff
@@ -29,46 +43,73 @@ def gpsdPoller(currentGPS):
...
+                epv = gpsd.fix.epv
+                ept = gpsd.fix.ept
+                gpt = gpsd.fix.time
                 #print 'time utc    ' , gpsd.utc,' + ', gpsd.fix.time
-                #print 'altitude (m)' , alt
                 currentGPS['lat'] = lat
                 currentGPS['lng'] = lng
                 currentGPS['alt'] = alt
+                log_message = "GPS: {}, {}, {}\n     {} epx:{} epv:{} ept:{}".format(lat, lng, alt, gpt, eph, epv, ept)
+                print log_message
```
Comment (rmspeers) [inline killerbee/zbwardrive/zbwardrive.py]: Prefer to let logging handle this -- at least if not in a verbose mode.

Comment (rfmcpherson) [inline killerbee/zbwardrive/zbwardrive.py]: The recent change should fix this.

**Comment by rfmcpherson** [comment]
Sorry. I want to get this PR in, but it's been so long since I looked at most of the code. I'm a little limited in my testing ability. I've got a Ravenstick, but no working GPS and there's not much IoT where I'm currently living.

**Comment by taylorcenters** [comment]
Closing - this issue has been idle for a long time. zbwardrive will be moving to a dedicated repo soon and this change will be re-explored

**Pull request closed without merging**